File: app/src/main/java/com/anbillon/lunarlite/app/LunarApplication.java
package com.anbillon.lunarlite.app;

import android.app.Application;
import com.anbillon.lunarlite.di.HasComponent;
import com.anbillon.lunarlite.di.component.AppComponent;
import com.anbillon.lunarlite.di.component.DaggerAppComponent;
import com.anbillon.lunarlite.di.module.AppModule;
import com.coolerfall.lunarlite.BuildConfig;
import timber.log.Timber;

/**
 * This is the lunar application through the whole project.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
public class LunarApplication extends Application implements HasComponent<AppComponent> {
	private static final String TAG = "LunarLite";
	private AppComponent appComponent;

	@Override public void onCreate() {
		super.onCreate();

		/* init injection */
		initInjection();

		/* config timber logger */
		Timber.tag(TAG);
		if (BuildConfig.DEBUG) {
			Timber.plant(new Timber.DebugTree());
		}
	}

	@Override public AppComponent getComponent() {
		return appComponent;
	}

	/* init dependency injection */
	private void initInjection() {
		appComponent = DaggerAppComponent.builder().appModule(new AppModule(this)).build();
	}
}

File: app/src/main/java/com/anbillon/lunarlite/app/LunarApplication.kt
package com.anbillon.lunarlite.app

import android.app.Activity
import android.app.Application
import com.anbillon.lunarlite.BuildConfig
import com.anbillon.lunarlite.di.component.DaggerAppComponent
import com.anbillon.lunarlite.di.module.AppModule
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasActivityInjector
import timber.log.Timber
import javax.inject.Inject

/**
 * This is the lunar application through the whole project.
 *
 * @author Vincent Cheung (coolingfall@gmail.com)
 */
class LunarApplication : Application(), HasActivityInjector {
  @Inject lateinit var activityInjector: DispatchingAndroidInjector<Activity>

  override fun onCreate() {
    super.onCreate()
    DaggerAppComponent.builder().appModule(AppModule(this)).build().inject(this)

    /* config timber logger */
    if (BuildConfig.DEBUG) {
      Timber.plant(Timber.DebugTree())
    }
  }

  override fun activityInjector(): AndroidInjector<Activity> {
    return activityInjector
  }
}
